### ISSUE TITLE
Guideline for static tables

### ISSUE THREAD

**Issue by 3lvis** [bakkenbaeck/iOS-playbook#72]
## Introduction

Let's say you have a static table view, where each cell needs a different treatment. One way to differentiate them is by assigning a type to the model so you can use a `switch` later. The problem is that initializing an enum for this `Type` later causes warnings in Xcode and doesn't compile.

In my project, I have a table with Album settings. This table has a few sections, one of this sections is "sharing".

<img width="411" alt="screen shot 2016-11-22 at 10 07 25" src="https://cloud.githubusercontent.com/assets/1088217/20517478/7d67a74e-b09b-11e6-9abc-6731fb11d840.png">

Here we have 3 rows. We need to detect taps in the first and third row, taps in the second row are ignored. We need to detect toggle switching in the second row. 

These are static cells, so we could use indexPath comparison to see if they are the first, the second or the third. But comparing to indexes can be difficult to read, that's why we have an enum to differentiate which type of row they are.

```swift
struct ShareSettingRow {
    enum `Type`: Int {
        case copySharingURL
        case anyoneCanAddPhotos
        case unshare

        var title: String {
            switch self {
            case .copySharingURL: return NSLocalizedString("Copy sharing URL", comment: "")
            case .anyoneCanAddPhotos: return NSLocalizedString("Anyone can add photos", comment: "")
            case .unshare: return NSLocalizedString("Unshare album", comment: "")
            }
        }
    }
}
```

These cells use the class `AlbumSettingCell`. Here's how we style them.

```swift
func update(albumSection: AlbumSettingsSection, album: Album, indexPath: IndexPath) {
    let sectionKind = albumSection.kind
    self.textLabel?.textColor = UIColor(hex: "6A6970")
    self.textLabel?.font = .light(16)
    self.detailTextLabel?.text = nil
    self.accessoryView = nil

    switch sectionKind {
    // ... more sections
    case .sharing:
        let sharingRowKind = ShareSettingRow.`Type`(rawValue: indexPath.row)!
        switch sharingRowKind {
        case .copySharingURL:
            self.textLabel?.textColor = UIColor(hex: "CC6EF1")
            self.textLabel?.font = .regular(16)
            self.accessoryView = UIImageView(image: UIImage(named: "get-link-gray")!)
        case .anyoneCanAddPhotos:
            self.accessoryView = self.switch
            break
        case .unshare:
            self.textLabel?.textColor = UIColor(hex: "EF3138")
            self.textLabel?.font = .regular(16)
        }

        self.textLabel?.text = sharingRowKind.title
    //... more sections
    }
}
```

## Problem

Swift doesn't let you do this:

```swift
let sharingRowKind = ShareSettingRow.`Type`(rawValue: indexPath.row)!
```

### Use Kind

Use Kind instead of `Type` for this cases since it's not a reserved word.

### Use models

Create a struct for each row and assign a type to them.

**Comment by elland** [comment]
When there are a multitude of possible solutions, why would we go for a workaround instead?

**Comment by 3lvis** [comment]
@Elland Let's find the best solution then.

**Comment by 3lvis** [comment]
I think using models here sounds like the best approach, also it allows us to move and add rows freely.

**Comment by elland** [comment]
I think the static table view should be treated no different than a dynamic one, the only difference being that the data source is immutable.

**Comment by 3lvis** [comment]
@Elland That still means we'll use `Type` here, though. I think you had an argument against it. Or have you changed your mind?

**Comment by elland** [comment]
No, it doesn't. This looks to me like a misuse of enums and structs. You're repurposing them by changing their type. They're here nothing more than a glorified dictionary, makes more sense to have a protocol they follow for this particular table view (title, image, maybe an action on selection or a view controller identifier it should use to push to if tapped?) and implement a simple struct that is just a glorified dictionary but that has hard-coded keys so there's auto-complete and no chance for typos etc.

**Comment by 3lvis** [comment]
@Elland We have to draw the over-engineering line somewhere, right?

**Comment by elland** [comment]
I don't follow.

**Comment by 3lvis** [comment]
@Elland You're proposing a complex solution for a very simple problem.

**Comment by elland** [comment]
Maybe you misread me, then? I was saying your solution for this problem was way too complex and looks a bit like misusing the API, by repurposing a struct by changing its type. It's not only over-engineered to solve a problem a dictionary can solve, but it's also hard to follow. Not only that but even the compiler started complaining about. 😄 

Another good hint that we're going the wrong way is that we can't find a way to name it.

**Comment by 3lvis** [comment]
Naming seems quite straight forward to me. Would be nice to see your solution, maybe I'm missing something :).

**Comment by elland** [comment]
I pasted some pseudocode on #ios yesterday, I believe.

**Comment by 3lvis** [comment]
Let's give it a try and convert it to sample code. Would be nice to reach an agreement into how to do this kind of things.

**Comment by elland** [comment]
I think this goes hand in hand with our table view/collection view model idea.

**Comment by 3lvis** [comment]
https://github.com/bakkenbaeck/Formulaic

**Comment by 3lvis** [comment]
Doing some spring cleaning, please open a new issue if you still want to do this.
